src/main/java/org/opensearch/security/user/UserService.java

```diff
@@ -0,0 +1,311 @@
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.security.DefaultObjectMapper;
+import org.opensearch.security.configuration.ConfigurationRepository;
+import org.opensearch.security.securityconf.DynamicConfigFactory;
+import org.opensearch.security.securityconf.Hashed;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.security.securityconf.impl.SecurityDynamicConfiguration;
+import org.opensearch.security.securityconf.impl.v7.InternalUserV7;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.support.SecurityJsonNode;
+
+import static org.opensearch.security.dlic.rest.support.Utils.hash;
+
+/**
+ * This class handles user registration and operations on behalf of the Security Plugin.
+ */
+public class UserService {
+
+    protected final Logger log = LogManager.getLogger(this.getClass());
+    ClusterService clusterService;
+    static ConfigurationRepository configurationRepository;
+    String securityIndex;
+    Client client;
+
+    final String NO_PASSWORD_OR_HASH_MESSAGE = "Please specify either 'hash' or 'password' when creating a new internal user.";
+    final String RESTRICTED_CHARACTER_USE_MESSAGE = "A restricted character(s) was detected in the account name. Please remove: ";
+
+    final String SERVICE_ACCOUNT_PASSWORD_MESSAGE = "A password cannot be provided for a service account. Failed to register service account: ";
+
+    final String SERVICE_ACCOUNT_HASH_MESSAGE = "A password hash cannot be provided for service account. Failed to register service account: ";
+
+    final String NO_ACCOUNT_NAME_MESSAGE = "No account name was specified in the request.";
+    protected static CType getConfigName() {
+        return CType.INTERNALUSERS;
+    }
+
+    static final List<String> RESTRICTED_FROM_USERNAME = ImmutableList.of(
+            ":" // Not allowed in basic auth, see https://stackoverflow.com/a/33391003/533057
+    );
+
+    // CreateUser
+    // Update User
+    // List User
+    // Get User
+    // This should be for all internal users instead of just service accounts
+
+
+    // Make a SecurityPluginUser object
+    // Work backwards from remaking this as a User Service -- can mostly reuse from current state
+    // Want to check that when code is created or updated, you check if it is a service account and turn on properties
+    // For example: Creator/owner of service account (extension or user)
+    // All end-to-end tests can be written without extensions and we can test everything
+    // Should not need to test create, update, list, get as unit tests but will want to test service account creation and deletion
+    // I.e. service account is invalid because no owner -- can leave tests till later.
```

Comment:
Not sure these comments are relevant, lets clean these up or promote to javadocs

Reply:
Oops